File: my-folder/0021-merge-two-sorted-lists/solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
var mergeTwoLists = function(list1, list2) {
    if (list1 === null && list2 === null) {
        return null;
    } else if (list1 === null) {
        return list2;
    } else if (list2 === null) {
        return list1;
    }
    let head = null;
    if (list1.val < list2.val) {
        head = list1;
        head.next = mergeTwoLists(list1.next, list2);
    } else {
        head = list2;
        head.next = mergeTwoLists(list1, list2.next);
    }

    return head;
};

File: my-folder/0021-merge-two-sorted-lists/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        merged_list: Optional[ListNode] = None
        cur: Optional[ListNode] = None

        while list1 and list2:
            if list1.val < list2.val:
                if not cur:
                    merged_list = cur = list1
                else:
                    cur.next = list1
                    cur = cur.next

                list1 = list1.next                
            else:
                if not cur:
                    merged_list = cur = list2
                else:
                    cur.next = list2
                    cur = cur.next
                list2 = list2.next

        if not list1 and not cur:
            merged_list = list2
        elif not list2 and not cur:
            merged_list = list1
        elif not list1:
            cur.next = list2
        else:
            cur.next = list1
        
        return merged_list
